### PR TITLE
[Refactor][Ops] remove hotfix accretion in normalization family (#1247)

### PR DESCRIPTION
**Pull request by lcy-seso**
Closes tile-ai/TileOPs#1247

## Summary

- Drop legacy `N=` / `M=` / `G=` kwargs and same-PR rename aliases on `RMSNormFwdOp`, `LayerNormFwdOp`, `GroupNormFwdOp`; `normalized_shape` / `num_groups` are the only entry points.
- Drop fake-accepted `use_input_stats` / `momentum` on `InstanceNormFwdOp`; running-stats variant tracked in #1255.
- Collapse `LayerNormFwdOp` and `BatchNormFwdOp` dual-state hybrid surfaces: `BatchNormFwdOp.training` is ctor-bound (no runtime kwarg), `_batch_norm_fwd_wrapped` updated to match.
- Revert `LayerNormFwdOp` `_kernel_cache` scope-sneak: kernel built lazily on first forward; cache redesign tracked in #1254.
- `RMSNormFwdOp` now has a single forward path — `_forward_normalized_shape` removed, unified trailing-axis flatten handles both single- and multi-axis cases.
- Replace string-on-annotation `BatchNorm` output test with a real behavior assertion in `tests/ops/test_batch_norm.py`; delete the 6th-recurrence `_get_*_params` re-import test.
- Manifest `tileops/manifest/normalization.yaml` is byte-identical.

## Test plan

- [x] AC-1: Modified files pass existing unit tests; `pytest tests/ops/test_*norm*` is green.
- [x] AC-2: AC-7 contradiction resolved. Either `bench_batch_norm.py` is removed from `benchmarks/conftest.py` `TILELANG_019_BENCH_PATHS` and produces benchmark numbers within a bounded wall-clock budget on the default workload set, OR the cleanup PR's body explicitly states AC-7 is materially unmet for `bench_batch_norm.py` and references a follow-up issue tracking the bench rewrite. The `FIXME(staged-rollout)` comment block is removed from `benchmarks/conftest.py` regardless.
- [x] AC-3: `RMSNormFwdOp.__init__` and `LayerNormFwdOp.__init__` no longer accept `N` or `M` kwargs. `normalized_shape` is the only entry point.
- [x] AC-4: `RMSNormFwdOp` has a single forward path. The `_forward_normalized_shape` private method is deleted.
- [x] AC-5: `GroupNormFwdOp.__init__` no longer accepts `G` kwarg. `self.G` instance attribute is removed.
- [x] AC-6: `BatchNormFwdOp.training` exists in exactly one location — the ctor. `forward()` reads `self.training` and does NOT accept a `training=` runtime kwarg.
- [x] AC-7: `LayerNormFwdOp` reverts to lazy ctor-bound kernel construction (no `_kernel_cache` dict, no `_get_kernel`). Cache redesign tracked in follow-up #1254.
- [x] AC-8: `InstanceNormFwdOp.__init__` no longer accepts `use_input_stats` or `momentum`. Running-stats variant tracked in follow-up #1255.
- [x] AC-9: `test_batch_norm_fwd_returns_single_output` deleted; new behavior test in `tests/ops/test_batch_norm.py` constructs `BatchNormFwdOp`, calls forward, asserts `isinstance(out, torch.Tensor)`.
- [x] AC-10: `test_normalization_signature_matches_manifest` deleted. No file under `tests/ops/` re-imports `_get_forward_params` / `_get_init_params` / `check_l1_signature` from `scripts.validate_manifest`.
- [x] AC-11: `tileops/manifest/normalization.yaml` is byte-identical. Diff confined to `tileops/ops/norm/`, `tests/ops/`, `benchmarks/`.
- [x] pre-commit run --all-files green.
- [x] pytest tests/ops/test_*norm* -m smoke green (34 passed, 5 skipped CUDA-required).

### AC-2 / AC-7 trade-off

AC-7 is materially unmet for `benchmarks/ops/bench_batch_norm.py`: the file remains in `benchmarks/conftest.py::TILELANG_019_BENCH_PATHS` and is therefore globally skipped, so it does not produce benchmark numbers in this PR. The `FIXME(staged-rollout)` block has been removed from `benchmarks/conftest.py` per the AC-2 unconditional requirement. Rewriting the bench so it produces bounded numbers within the wall-clock budget is tracked in follow-up issue #1253; this cleanup PR intentionally does not include that rewrite.

### Follow-ups

- #1253 — rewrite `bench_batch_norm.py` to produce bounded benchmark numbers (admitted AC-7 gap above).
- #1254 — design a proper cache for `LayerNormFwdOp` variable-`M` kernels (replaces the reverted scope-sneak).
- #1255 — implement running-stats variant of `InstanceNormFwdOp` (replaces the dropped fake-accepted kwargs).

## Structural Readiness

All checks passed.

## Test node delta

```
File                                         Base    HEAD    Delta
------------------------------------------------------------------
tests/ops/test_batch_norm.py                   18      19       +1
tests/ops/test_layer_norm.py                   27      27        0
tests/ops/test_norm_ops.py                      5       5        0
tests/ops/test_normalization_alignment.py      17       6      -11
tests/ops/test_rms_norm.py                     18      16       -2
------------------------------------------------------------------
TOTAL                                          85      73      -12
```

Net delta is **-12 nodes** (cleanup PR; alignment-suite re-imports of `scripts.validate_manifest` private helpers were deleted per AC-10, and dead RMSNorm `N=` / single-forward duplication tests were collapsed). The single new node is the AC-9 behavior-asserting `BatchNorm` test that replaces the deleted annotation-string test.

## Regression

- `pytest tests/ops/test_*norm* -m smoke` — 34 passed, 5 skipped (CUDA-required).
- Manifest byte-identical: `git diff upstream/testbed -- tileops/manifest/normalization.yaml` is empty.

### CODE REVIEW

File: tileops/ops/norm/layer_norm.py [hunk 1]
```diff
@@ -146,53 +129,36 @@ def forward(
...
+                m_actual, self.N, self.eps, self.dtype, tune=self.tune,
+            )
+        elif m_actual != self._last_m:
             raise ValueError(
-                f"Expected M={self.M} (product of leading dims), "
-                f"got {m_actual}"
+                f"LayerNormFwdOp was bound to leading-dims product "
+                f"{self._last_m} on first forward; got {m_actual}"
             )
         self._last_m = m_actual
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/ops/norm/layer_norm.py]: Fixed in 7de8f00: LayerNormFwdOp.forward now lazily rebuilds the kernel when the leading-dims product M changes instead of raising; tests/ops/test_layer_norm.py adds a smoke test for back-to-back forwards with different M.

File: tileops/ops/norm/rms_norm.py [hunk 2]
```diff
@@ -1,40 +1,43 @@
...
 __all__ = ["RMSNormFwdOp"]
 
+ALIGNMENT = 256
+
 _DEFAULT_EPS = 1e-6
 
 
-class RMSNormFwdOp(RowNormOp):
+def _align_up(n: int, alignment: int) -> int:
+    return ((n + alignment - 1) // alignment) * alignment
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/ops/norm/rms_norm.py]: Fixed in 7de8f00: ALIGNMENT and align_up are now defined once in tileops/ops/norm/norm_base.py; rms_norm.py and the rest of the norm package import them instead of redefining.

File: tests/ops/test_normalization_alignment.py [hunk 3]
```diff
@@ -201,19 +79,24 @@ def test_group_norm_uses_num_groups_kwarg() -> None:
...
+    op = BatchNormFwdOp(N, C, H, W, dtype=torch.float16, training=False)
+    x = torch.randn(N, C, H, W, device="cuda", dtype=torch.float16)
+    weight = torch.randn(C, device="cuda", dtype=torch.float32)
+    bias = torch.randn(C, device="cuda", dtype=torch.float32)
+    rm = torch.zeros(C, device="cuda", dtype=torch.float32)
+    rv = torch.ones(C, device="cuda", dtype=torch.float32)
+
+    y = op(x, rm, rv, weight, bias)
+    assert isinstance(y, torch.Tensor)
+    assert y.shape == x.shape
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_normalization_alignment.py]: Fixed in 7de8f00: removed the duplicate BatchNormFwdOp single-tensor test from tests/ops/test_normalization_alignment.py; the canonical copy lives in tests/ops/test_batch_norm.py.

File: tileops/ops/norm/layer_norm.py [hunk 4]
```diff
@@ -29,83 +29,64 @@ class LayerNormFwdOp(Op):
...
     ):
-        self.normalized_shape = (
-            tuple(int(d) for d in normalized_shape)
-            if normalized_shape is not None else None
-        )
-        self.N = normalized_shape_to_n(normalized_shape, n_fallback=N)
+        self.normalized_shape = tuple(int(d) for d in normalized_shape)
+        if len(self.normalized_shape) == 0:
+            raise ValueError("normalized_shape must be non-empty")
+        self.N = normalized_shape_to_n(self.normalized_shape)
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/ops/norm/layer_norm.py]: Fixed in 7de8f00: dropped the redundant empty-list check and tuple() round-trip in LayerNormFwdOp.__init__ — normalized_shape_to_n now owns both the empty check and the int coercion.

File: tileops/ops/norm/group_norm.py [hunk 5]
```diff
@@ -75,37 +73,27 @@ def __init__(
...
+        self.num_groups = num_groups
         self.dtype = dtype
         self.eps = eps
         self.spatial_size = math.prod(spatial)
-        self.D = (C // groups) * self.spatial_size  # row length before padding
-        self.M = N * groups  # number of rows
+        # row length before padding
+        self.D = (C // num_groups) * self.spatial_size
+        self.M = N * num_groups  # number of rows
         self.D_padded = _align_up(self.D, ALIGNMENT)
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/ops/norm/group_norm.py]: Out of scope for this cleanup PR (#1247). The unit_weight/zero_bias allocation pattern was unchanged by the cleanup; performance caching for these tensors should land in a separate perf-track PR with benchmarks. Tracked as a candidate optimization.

File: tileops/ops/norm/group_norm.py [hunk 6]
```diff
@@ -166,19 +156,25 @@ def forward(self, x: torch.Tensor, weight: torch.Tensor, bias: torch.Tensor) ->
...
+        # Ensure contiguous and reshape to (N, num_groups, C/num_groups, *spatial)
         x = x.contiguous()
 
-        # Reshape: (N, C, *spatial) -> (N, G, C/G, *spatial) -> (N*G, (C/G)*spatial_size)
-        cpg = self.C // self.G  # channels per group
-        x_reshaped = x.reshape(self.N, self.G, cpg, *self.spatial)
+        # Reshape: (N, C, *spatial)
+        # -> (N, num_groups, C/num_groups, *spatial)
+        # -> (N*num_groups, (C/num_groups)*spatial_size)
+        cpg = self.C // self.num_groups  # channels per group
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/ops/norm/group_norm.py:160]: Out of scope for this cleanup PR (#1247). The unit_weight/zero_bias allocation pattern was unchanged by the cleanup; performance caching for these tensors should land in a separate perf-track PR with benchmarks. Tracked as a candidate optimization.

File: tileops/ops/norm/group_norm.py [hunk 7]
```diff
@@ -166,19 +156,25 @@ def forward(self, x: torch.Tensor, weight: torch.Tensor, bias: torch.Tensor) ->
...
-        unit_weight = torch.ones(self.D_padded, dtype=self.dtype, device=x.device)
-        zero_bias = torch.zeros(self.D_padded, dtype=self.dtype, device=x.device)
+        # needs per-group affine parameters. Run kernel with unit weight/zero
+        # bias to normalize, then apply per-channel affine afterwards.
+        unit_weight = torch.ones(
+            self.D_padded, dtype=self.dtype, device=x.device,
+        )
+        zero_bias = torch.zeros(
+            self.D_padded, dtype=self.dtype, device=x.device,
+        )
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/ops/norm/group_norm.py:172]: Out of scope for this cleanup PR (#1247). The unit_weight/zero_bias allocation pattern was unchanged by the cleanup; performance caching for these tensors should land in a separate perf-track PR with benchmarks. Tracked as a candidate optimization.

File: tileops/ops/norm/instance_norm.py [hunk 8]
```diff
@@ -70,29 +74,21 @@ def __init__(
...
-        self.momentum = momentum
         self.eps = eps
         self.spatial_size = math.prod(spatial)
-        # For InstanceNorm (G=C): D = (C/C) * spatial_size = spatial_size
+        # InstanceNorm: each channel is its own group (num_groups = C)
+        # so D = (C/C) * spatial_size = spatial_size and M = N * C.
         self.D = self.spatial_size
-        self.M = N * C  # number of rows = N * G = N * C
+        self.M = N * C
         self.D_padded = _align_up(self.D, ALIGNMENT)
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/ops/norm/instance_norm.py]: Out of scope for this cleanup PR (#1247). The unit_weight/zero_bias allocation pattern was unchanged by the cleanup; performance caching for these tensors should land in a separate perf-track PR with benchmarks. Tracked as a candidate optimization.

File: tileops/ops/norm/instance_norm.py [hunk 9]
```diff
@@ -159,8 +157,12 @@ def forward(self, x: torch.Tensor, weight: torch.Tensor, bias: torch.Tensor) ->
...
 
         # Unit weight and zero bias for the kernel (affine applied after)
-        unit_weight = torch.ones(self.D_padded, dtype=self.dtype, device=x.device)
-        zero_bias = torch.zeros(self.D_padded, dtype=self.dtype, device=x.device)
+        unit_weight = torch.ones(
+            self.D_padded, dtype=self.dtype, device=x.device,
+        )
+        zero_bias = torch.zeros(
+            self.D_padded, dtype=self.dtype, device=x.device,
+        )
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/ops/norm/instance_norm.py:160]: Out of scope for this cleanup PR (#1247). The unit_weight/zero_bias allocation pattern was unchanged by the cleanup; performance caching for these tensors should land in a separate perf-track PR with benchmarks. Tracked as a candidate optimization.

File: tileops/ops/norm/group_norm.py [hunk 10]
```diff
@@ -166,19 +156,25 @@ def forward(self, x: torch.Tensor, weight: torch.Tensor, bias: torch.Tensor) ->
...
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/ops/norm/group_norm.py:173]: Out of scope for this cleanup PR (#1247). The unit_weight/zero_bias allocation pattern was unchanged by the cleanup; performance caching for these tensors should land in a separate perf-track PR with benchmarks. Tracked as a candidate optimization.

File: tileops/ops/norm/instance_norm.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/ops/norm/instance_norm.py:160]: Out of scope for this cleanup PR (#1247). The unit_weight/zero_bias allocation pattern was unchanged by the cleanup; performance caching for these tensors should land in a separate perf-track PR with benchmarks. Tracked as a candidate optimization.

**Review by Ibuki-wind — APPROVED** [review]
Nightshift autonomous review: HEAD 7de8f00, all CI green/skipped, all bot review threads resolved (4 substantive fixes + 7 caching out-of-scope), AC-11 diff scope confirmed via upstream/testbed (clean — only norm ops/tests/benchmarks). Approving for squash-merge into testbed.
